File: brique-35-sdk-auth/android/src/main/java/com/molam/id/MolamIdClient.kt
package com.molam.id

import android.content.Context
import android.os.Build
import androidx.work.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.*
import java.util.concurrent.TimeUnit

class MolamIdClient(
    private val context: Context,
    private val baseUrl: String,
    private val heartbeatIntervalMin: Long = 15
) {
    private val store: TokenStore = SecureTokenStore(context)
    private val client = OkHttpClient.Builder()
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .connectTimeout(30, TimeUnit.SECONDS)
        .build()

    // Callbacks
    var onAnomaly: ((kind: String, severity: String, details: Any?) -> Unit)? = null
    var onTokenUpdate: ((tokens: Tokens?) -> Unit)? = null
    var onError: ((error: Exception) -> Unit)? = null

    // Metrics
    private var metrics = Metrics()

    data class Metrics(
        var loginAttempts: Int = 0,
        var refreshes: Int = 0,
        var heartbeatsSent: Int = 0,
        var revocations: Int = 0
    )

    // ============================================================================
    // AUTHENTICATION
    // ============================================================================

    suspend fun signup(
        email: String? = null,
        phone: String? = null,
        password: String,
        locale: String? = null,
        country: String? = null
    ): JSONObject = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            email?.let { put("email", it) }
            phone?.let { put("phone", it) }
            put("password", password)
            locale?.let { put("locale", it) }
            country?.let { put("country", it) }
        }

        post("/api/id/auth/signup", body)
    }

    suspend fun login(identifier: String, password: String): Tokens = withContext(Dispatchers.IO) {
        metrics.loginAttempts++

        val body = JSONObject().apply {
            put("identifier", identifier)
            put("password", password)
            put("device", devicePayload())
        }

        val response = post("/api/id/auth/login", body)
        val tokens = adaptTokens(response)

        store.save(tokens)
        onTokenUpdate?.invoke(tokens)
        scheduleHeartbeat()

        tokens
    }

    suspend fun refresh(): Tokens = withContext(Dispatchers.IO) {
        metrics.refreshes++

        val currentTokens = store.load() ?: throw MolamIdException("not_authenticated")

        val body = JSONObject().apply {
            put("refresh_token", currentTokens.refreshToken)
        }

        val response = post("/api/id/auth/refresh", body)
        val newTokens = Tokens(
            accessToken = response.getString("access_token"),
            refreshToken = response.getString("refresh_token"),
            expiresAt = System.currentTimeMillis() + response.optLong("expires_in", 900) * 1000,
            sessionId = currentTokens.sessionId
        )

        store.save(newTokens)
        onTokenUpdate?.invoke(newTokens)

        newTokens
    }

    suspend fun logout(revokeAll: Boolean = false): Unit = withContext(Dispatchers.IO) {
        val tokens = store.load() ?: return@withContext

        try {
            val path = if (revokeAll) "/api/id/sessions/revoke_all" else "/api/id/sessions/${tokens.sessionId}/revoke"
            postAuth(path, JSONObject())
            metrics.revocations++
        } catch (e: Exception) {
            // Ignore errors during logout
        }

        cancelHeartbeat()
        store.save(null)
        onTokenUpdate?.invoke(null)
    }

    // ============================================================================
    // SESSION MANAGEMENT
    // ============================================================================

    suspend fun mySessions(): List<JSONObject> = withContext(Dispatchers.IO) {
        val response = getAuth("/api/id/sessions/me")
        val sessions = response.optJSONArray("sessions") ?: return@withContext emptyList()

        List(sessions.length()) { i -> sessions.getJSONObject(i) }
    }

    suspend fun revokeSession(sessionId: String): Unit = withContext(Dispatchers.IO) {
        postAuth("/api/id/sessions/$sessionId/revoke", JSONObject())
        metrics.revocations++
    }

    suspend fun revokeAllSessions(): Unit = withContext(Dispatchers.IO) {
        postAuth("/api/id/sessions/revoke_all", JSONObject())
        metrics.revocations++
    }

    // ============================================================================
    // HEARTBEAT
    // ============================================================================

    private fun scheduleHeartbeat() {
        val workRequest = PeriodicWorkRequestBuilder<HeartbeatWorker>(
            heartbeatIntervalMin,
            TimeUnit.MINUTES
        )
            .setConstraints(
                Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.CONNECTED)
                    .build()
            )
            .addTag(TAG_HEARTBEAT)
            .build()

        WorkManager.getInstance(context).enqueueUniquePeriodicWork(
            WORK_HEARTBEAT,
            ExistingPeriodicWorkPolicy.UPDATE,
            workRequest
        )
    }

    private fun cancelHeartbeat() {
        WorkManager.getInstance(context).cancelUniqueWork(WORK_HEARTBEAT)
    }

    internal suspend fun heartbeat() = withContext(Dispatchers.IO) {
        val tokens = ensureFreshToken()
        val body = JSONObject().apply {
            put("session_id", tokens.sessionId)
        }

        postAuth("/api/id/sessions/heartbeat", body)
        metrics.heartbeatsSent++
    }

    // ============================================================================
    // USSD
    // ============================================================================

    suspend fun ussdBind(msisdn: String, countryCode: String? = null): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("msisdn", msisdn)
            put("country_code", countryCode ?: "auto")
        }

        postAuth("/api/id/ussd/bind", body)
    }

    // ============================================================================
    // USER INFO
    // ============================================================================

    suspend fun isAuthenticated(): Boolean {
        val tokens = store.load() ?: return false

        return if (tokens.isExpired()) {
            try {
                refresh()
                true
            } catch (e: Exception) {
                false
            }
        } else {
            true
        }
    }

    fun getMetrics(): Map<String, Int> = mapOf(
        "loginAttempts" to metrics.loginAttempts,
        "refreshes" to metrics.refreshes,
        "heartbeatsSent" to metrics.heartbeatsSent,
        "revocations" to metrics.revocations
    )

    // ============================================================================
    // HELPERS
    // ============================================================================

    private fun devicePayload(): JSONObject = JSONObject().apply {
        put("fingerprint", JSONObject().apply {
            put("brand", Build.BRAND)
            put("model", Build.MODEL)
            put("sdk", Build.VERSION.SDK_INT)
            put("tz", TimeZone.getDefault().id)
        })
    }

    private fun adaptTokens(response: JSONObject): Tokens {
        val expiresIn = response.optLong("expires_in", 900)
        return Tokens(
            accessToken = response.getString("access_token"),
            refreshToken = response.getString("refresh_token"),
            expiresAt = System.currentTimeMillis() + expiresIn * 1000,
            sessionId = response.getString("session_id")
        )
    }

    private suspend fun ensureFreshToken(): Tokens {
        var tokens = store.load() ?: throw MolamIdException("not_authenticated")

        val skew = 30_000L // 30 seconds
        if (System.currentTimeMillis() >= tokens.expiresAt - skew) {
            tokens = refresh()
        }

        return tokens
    }

    // ============================================================================
    // HTTP METHODS
    // ============================================================================

    private fun getAuth(path: String): JSONObject {
        val tokens = store.load() ?: throw MolamIdException("not_authenticated")

        val request = Request.Builder()
            .url("$baseUrl$path")
            .get()
            .addHeader("Authorization", "Bearer ${tokens.accessToken}")
            .build()

        return execute(request)
    }

    private fun post(path: String, body: JSONObject): JSONObject {
        val request = Request.Builder()
            .url("$baseUrl$path")
            .post(body.toString().toRequestBody(MEDIA_TYPE_JSON))
            .addHeader("Content-Type", "application/json")
            .build()

        return execute(request)
    }

    private fun postAuth(path: String, body: JSONObject): JSONObject {
        val tokens = store.load() ?: throw MolamIdException("not_authenticated")

        val request = Request.Builder()
            .url("$baseUrl$path")
            .post(body.toString().toRequestBody(MEDIA_TYPE_JSON))
            .addHeader("Content-Type", "application/json")
            .addHeader("Authorization", "Bearer ${tokens.accessToken}")
            .build()

        return execute(request)
    }

    private fun execute(request: Request): JSONObject {
        client.newCall(request).execute().use { response ->
            // Check for rate limiting
            if (response.code == 429) {
                throw MolamIdException("rate_limited")
            }

            // Check for anomaly header
            response.header("x-molam-anomaly")?.let { anomalyHeader ->
                try {
                    val anomaly = JSONObject(anomalyHeader)
                    onAnomaly?.invoke(
                        kind = anomaly.getString("kind"),
                        severity = anomaly.getString("severity"),
                        details = anomaly.opt("details")
                    )
                } catch (e: Exception) {
                    // Ignore parse errors
                }
            }

            // Handle success
            if (response.isSuccessful) {
                val bodyString = response.body?.string() ?: "{}"
                return if (bodyString.isBlank()) JSONObject() else JSONObject(bodyString)
            }

            // Handle authentication errors
            if (response.code == 401) {
                store.save(null)
                onTokenUpdate?.invoke(null)
                cancelHeartbeat()
                throw MolamIdException("unauthorized")
            }

            throw MolamIdException("http_${response.code}: ${response.body?.string()}")
        }
    }

    companion object {
        private val MEDIA_TYPE_JSON = "application/json; charset=utf-8".toMediaType()
        private const val TAG_HEARTBEAT = "molam_id_heartbeat"
        private const val WORK_HEARTBEAT = "molam_id_heartbeat_work"
    }
}

// ============================================================================
// EXCEPTIONS
// ============================================================================

class MolamIdException(message: String) : Exception(message)

// ============================================================================
// HEARTBEAT WORKER
// ============================================================================

class HeartbeatWorker(
    context: Context,
    params: WorkerParameters
) : CoroutineWorker(context, params) {

    override suspend fun doWork(): Result {
        // Note: In production, you'd need to pass baseUrl from WorkData
        // or store it in SharedPreferences
        return try {
            val store = SecureTokenStore(applicationContext)
            val tokens = store.load() ?: return Result.success()

            // Simplified heartbeat - in production, use full client
            val client = OkHttpClient()
            val body = JSONObject().apply {
                put("session_id", tokens.sessionId)
            }.toString()

            val request = Request.Builder()
                .url("${inputData.getString("baseUrl")}/api/id/sessions/heartbeat")
                .post(body.toRequestBody("application/json".toMediaType()))
                .addHeader("Authorization", "Bearer ${tokens.accessToken}")
                .build()

            client.newCall(request).execute().close()
            Result.success()
        } catch (e: Exception) {
            Result.retry()
        }
    }
}

File: brique-35-sdk-auth/ios/Sources/MolamID/MolamIdClient.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif
#if canImport(AppKit)
import AppKit
#endif

public final class MolamIdClient {
    private let baseURL: URL
    private let store: TokenStore
    private var heartbeatTimer: Timer?
    private let heartbeatInterval: TimeInterval

    // Callbacks
    public var onAnomaly: ((AnomalyEvent) -> Void)?
    public var onTokenUpdate: ((Tokens?) -> Void)?
    public var onError: ((Error) -> Void)?

    // Metrics
    private var metrics = Metrics()

    struct Metrics {
        var loginAttempts = 0
        var refreshes = 0
        var heartbeatsSent = 0
        var revocations = 0
    }

    // MARK: - Initialization

    public init(baseUrl: String, heartbeatIntervalSec: TimeInterval = 120) {
        self.baseURL = URL(string: baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")))!
        self.store = KeychainTokenStore()
        self.heartbeatInterval = heartbeatIntervalSec
    }

    // MARK: - Authentication

    public func login(identifier: String, password: String) async throws -> Tokens {
        metrics.loginAttempts += 1

        let body: [String: Any] = [
            "identifier": identifier,
            "password": password,
            "device": devicePayload()
        ]

        let response = try await post("/api/id/auth/login", body: body)
        let tokens = adaptTokens(response)

        store.save(tokens)
        onTokenUpdate?(tokens)
        startHeartbeat()

        return tokens
    }

    public func signup(email: String?, phone: String?, password: String, locale: String? = nil, country: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = ["password": password]
        if let email = email { body["email"] = email }
        if let phone = phone { body["phone"] = phone }
        if let locale = locale { body["locale"] = locale }
        if let country = country { body["country"] = country }

        return try await post("/api/id/auth/signup", body: body)
    }

    public func refresh() async throws -> Tokens {
        metrics.refreshes += 1

        guard let currentTokens = store.load() else {
            throw MolamIdError.notAuthenticated
        }

        let body: [String: Any] = ["refresh_token": currentTokens.refreshToken]
        let response = try await post("/api/id/auth/refresh", body: body)
        let newTokens = adaptTokens(response, sessionId: currentTokens.sessionId)

        store.save(newTokens)
        onTokenUpdate?(newTokens)

        return newTokens
    }

    public func logout(revokeAll: Bool = false) async {
        guard let tokens = store.load() else { return }

        let path = revokeAll ? "/api/id/sessions/revoke_all" : "/api/id/sessions/\(tokens.sessionId)/revoke"
        _ = try? await postAuth(path, body: [:])

        metrics.revocations += 1
        stopHeartbeat()
        store.save(nil)
        onTokenUpdate?(nil)
    }

    // MARK: - Session Management

    public func mySessions() async throws -> [SessionInfo] {
        let response = try await getAuth("/api/id/sessions/me")
        guard let sessionsDict = response["sessions"] as? [[String: Any]] else {
            return []
        }

        return sessionsDict.compactMap { dict in
            guard let data = try? JSONSerialization.data(withJSONObject: dict),
                  let session = try? JSONDecoder().decode(SessionInfo.self, from: data) else {
                return nil
            }
            return session
        }
    }

    public func revokeSession(_ sessionId: String) async throws {
        _ = try await postAuth("/api/id/sessions/\(sessionId)/revoke", body: [:])
        metrics.revocations += 1
    }

    public func revokeAllSessions() async throws {
        _ = try await postAuth("/api/id/sessions/revoke_all", body: [:])
        metrics.revocations += 1
    }

    // MARK: - Heartbeat

    private func startHeartbeat() {
        stopHeartbeat()
        heartbeatTimer = Timer.scheduledTimer(withTimeInterval: heartbeatInterval, repeats: true) { [weak self] _ in
            Task {
                try? await self?.heartbeat()
            }
        }
    }

    private func stopHeartbeat() {
        heartbeatTimer?.invalidate()
        heartbeatTimer = nil
    }

    public func heartbeat() async throws {
        guard let tokens = try await ensureFreshToken() else { return }

        let body: [String: Any] = ["session_id": tokens.sessionId]
        _ = try await postAuth("/api/id/sessions/heartbeat", body: body)

        metrics.heartbeatsSent += 1
    }

    // MARK: - USSD

    public func ussdBind(msisdn: String, countryCode: String? = nil) async throws {
        let body: [String: Any] = [
            "msisdn": msisdn,
            "country_code": countryCode ?? "auto"
        ]
        _ = try await postAuth("/api/id/ussd/bind", body: body)
    }

    // MARK: - User Info

    public func isAuthenticated() async -> Bool {
        guard let tokens = store.load() else { return false }

        if tokens.isExpired {
            do {
                _ = try await refresh()
                return true
            } catch {
                return false
            }
        }

        return true
    }

    public func getMetrics() -> [String: Int] {
        return [
            "loginAttempts": metrics.loginAttempts,
            "refreshes": metrics.refreshes,
            "heartbeatsSent": metrics.heartbeatsSent,
            "revocations": metrics.revocations
        ]
    }

    // MARK: - Helpers

    private func adaptTokens(_ response: [String: Any], sessionId: String? = nil) -> Tokens {
        let expiresIn = (response["expires_in"] as? TimeInterval) ?? 900
        let now = Date().timeIntervalSince1970 * 1000

        return Tokens(
            accessToken: response["access_token"] as! String,
            refreshToken: response["refresh_token"] as! String,
            expiresAt: now + (expiresIn * 1000),
            sessionId: sessionId ?? (response["session_id"] as! String)
        )
    }

    private func ensureFreshToken() async throws -> Tokens? {
        guard var tokens = store.load() else { return nil }

        let skew: TimeInterval = 30_000 // 30 seconds
        if Date().timeIntervalSince1970 * 1000 >= tokens.expiresAt - skew {
            tokens = try await refresh()
        }

        return tokens
    }

    private func devicePayload() -> [String: Any] {
        var fingerprint: [String: Any] = [:]

        #if canImport(UIKit)
        let device = UIDevice.current
        fingerprint["model"] = device.model
        fingerprint["systemName"] = device.systemName
        fingerprint["systemVersion"] = device.systemVersion
        #elseif canImport(AppKit)
        fingerprint["model"] = "Mac"
        fingerprint["systemName"] = "macOS"
        #endif

        fingerprint["tz"] = TimeZone.current.identifier

        return ["fingerprint": fingerprint]
    }

    // MARK: - HTTP Methods

    private func getAuth(_ path: String) async throws -> [String: Any] {
        guard let tokens = try await ensureFreshToken() else {
            throw MolamIdError.notAuthenticated
        }

        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.setValue("Bearer \(tokens.accessToken)", forHTTPHeaderField: "Authorization")

        return try await execute(request)
    }

    private func post(_ path: String, body: [String: Any]) async throws -> [String: Any] {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        return try await execute(request)
    }

    private func postAuth(_ path: String, body: [String: Any]) async throws -> [String: Any] {
        guard let tokens = try await ensureFreshToken() else {
            throw MolamIdError.notAuthenticated
        }

        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(tokens.accessToken)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        return try await execute(request)
    }

    private func execute(_ request: URLRequest) async throws -> [String: Any] {
        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw MolamIdError.invalidResponse
        }

        // Handle rate limiting
        if httpResponse.statusCode == 429 {
            throw MolamIdError.rateLimited
        }

        // Handle anomaly header
        if let anomalyHeader = httpResponse.value(forHTTPHeaderField: "x-molam-anomaly"),
           let anomalyData = anomalyHeader.data(using: .utf8),
           let anomaly = try? JSONDecoder().decode(AnomalyEvent.self, from: anomalyData) {
            onAnomaly?(anomaly)
        }

        // Check for success
        guard (200..<300).contains(httpResponse.statusCode) else {
            // Handle auth errors
            if httpResponse.statusCode == 401 {
                store.save(nil)
                onTokenUpdate?(nil)
                stopHeartbeat()
                throw MolamIdError.unauthorized
            }

            throw MolamIdError.httpError(httpResponse.statusCode)
        }

        if data.isEmpty {
            return [:]
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return [:]
        }

        return json
    }
}

// MARK: - Errors

public enum MolamIdError: Error {
    case notAuthenticated
    case invalidResponse
    case rateLimited
    case unauthorized
    case httpError(Int)
}
